File: backend-java/monolith/src/main/java/com/freelms/lms/payment/service/PaymentService.java
package com.freelms.lms.payment.service;

import com.freelms.lms.common.dto.PagedResponse;
import com.freelms.lms.common.enums.PaymentStatus;
import com.freelms.lms.common.exception.BadRequestException;
import com.freelms.lms.common.exception.ResourceNotFoundException;
import com.freelms.lms.payment.dto.CreatePaymentRequest;
import com.freelms.lms.payment.dto.PaymentDto;
import com.freelms.lms.payment.entity.Payment;
import com.freelms.lms.payment.repository.PaymentRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Slf4j
@Service
@RequiredArgsConstructor
public class PaymentService {

    private final PaymentRepository paymentRepository;

    @Transactional
    public PaymentDto createPayment(Long userId, CreatePaymentRequest request) {
        log.info("Creating payment for user {} course {}", userId, request.getCourseId());

        if (paymentRepository.existsByUserIdAndCourseIdAndStatus(userId, request.getCourseId(), PaymentStatus.COMPLETED)) {
            throw new BadRequestException("Payment already completed for this course");
        }

        Payment payment = Payment.builder()
                .userId(userId)
                .courseId(request.getCourseId())
                .amount(request.getAmount())
                .currency(request.getCurrency())
                .paymentGateway(request.getPaymentGateway())
                .status(PaymentStatus.PENDING)
                .build();

        payment = paymentRepository.save(payment);
        log.info("Payment created with ID: {}", payment.getId());

        return toDto(payment);
    }

    @Transactional
    public PaymentDto completePayment(Long paymentId, String transactionId) {
        Payment payment = paymentRepository.findById(paymentId)
                .orElseThrow(() -> new ResourceNotFoundException("Payment", "id", paymentId));

        if (payment.getStatus() != PaymentStatus.PENDING) {
            throw new BadRequestException("Payment is not in pending status");
        }

        payment.complete(transactionId != null ? transactionId : UUID.randomUUID().toString());
        payment = paymentRepository.save(payment);

        log.info("Payment completed: {}", paymentId);
        return toDto(payment);
    }

    @Transactional
    public PaymentDto failPayment(Long paymentId, String reason) {
        Payment payment = paymentRepository.findById(paymentId)
                .orElseThrow(() -> new ResourceNotFoundException("Payment", "id", paymentId));

        payment.fail(reason);
        payment = paymentRepository.save(payment);

        log.info("Payment failed: {} - {}", paymentId, reason);
        return toDto(payment);
    }

    @Transactional
    public PaymentDto refundPayment(Long paymentId, String reason) {
        Payment payment = paymentRepository.findById(paymentId)
                .orElseThrow(() -> new ResourceNotFoundException("Payment", "id", paymentId));

        if (payment.getStatus() != PaymentStatus.COMPLETED) {
            throw new BadRequestException("Only completed payments can be refunded");
        }

        payment.refund(reason);
        payment = paymentRepository.save(payment);

        log.info("Payment refunded: {} - {}", paymentId, reason);
        return toDto(payment);
    }

    @Transactional(readOnly = true)
    public PaymentDto getPaymentById(Long id) {
        Payment payment = paymentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Payment", "id", id));
        return toDto(payment);
    }

    @Transactional(readOnly = true)
    public PagedResponse<PaymentDto> getUserPayments(Long userId, Pageable pageable) {
        Page<Payment> payments = paymentRepository.findByUserId(userId, pageable);
        List<PaymentDto> dtos = payments.getContent().stream().map(this::toDto).toList();
        return PagedResponse.of(payments, dtos);
    }

    @Transactional(readOnly = true)
    public boolean hasCompletedPayment(Long userId, Long courseId) {
        return paymentRepository.existsByUserIdAndCourseIdAndStatus(userId, courseId, PaymentStatus.COMPLETED);
    }

    private PaymentDto toDto(Payment payment) {
        return PaymentDto.builder()
                .id(payment.getId())
                .userId(payment.getUserId())
                .courseId(payment.getCourseId())
                .amount(payment.getAmount())
                .currency(payment.getCurrency())
                .status(payment.getStatus())
                .paymentGateway(payment.getPaymentGateway())
                .transactionId(payment.getTransactionId())
                .processedAt(payment.getProcessedAt())
                .createdAt(payment.getCreatedAt())
                .build();
    }
}

File: frontend/src/app/core/services/payment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { environment } from '@environments/environment';
import {
  ApiResponse,
  PagedResponse,
  Payment,
  CreatePaymentRequest
} from '../models';

@Injectable({
  providedIn: 'root'
})
export class PaymentService {
  private readonly apiUrl = `${environment.apiUrl}/payments`;

  constructor(private http: HttpClient) {}

  createPayment(request: CreatePaymentRequest): Observable<Payment> {
    return this.http.post<ApiResponse<Payment>>(this.apiUrl, request)
      .pipe(map(res => res.data));
  }

  getPaymentById(id: number): Observable<Payment> {
    return this.http.get<ApiResponse<Payment>>(`${this.apiUrl}/${id}`)
      .pipe(map(res => res.data));
  }

  getMyPayments(page = 0, size = 20): Observable<PagedResponse<Payment>> {
    const params = new HttpParams()
      .set('page', page.toString())
      .set('size', size.toString());

    return this.http.get<ApiResponse<PagedResponse<Payment>>>(`${this.apiUrl}/my`, { params })
      .pipe(map(res => res.data));
  }

  completePayment(id: number, transactionId?: string): Observable<Payment> {
    let params = new HttpParams();
    if (transactionId) {
      params = params.set('transactionId', transactionId);
    }

    return this.http.post<ApiResponse<Payment>>(`${this.apiUrl}/${id}/complete`, null, { params })
      .pipe(map(res => res.data));
  }

  refundPayment(id: number, reason: string): Observable<Payment> {
    const params = new HttpParams().set('reason', reason);
    return this.http.post<ApiResponse<Payment>>(`${this.apiUrl}/${id}/refund`, null, { params })
      .pipe(map(res => res.data));
  }
}
